lib/code_corps/model/project.ex

```diff
@@ -71,12 +74,69 @@ defmodule CodeCorps.Project do
     struct
     |> changeset(params)
     |> cast(params, [:organization_id])
+    |> validate_required([:cloudinary_public_id, :description])
+    |> associate_categories(params)
+    |> validate_length(:categories, min: 1)
+    |> associate_skills(params)
+    |> validate_length(:skills, min: 1)
     |> put_assoc(:task_lists, TaskList.default_task_lists())
     |> put_member_assoc()
     |> generate_icon_color(:default_color)
     |> assoc_constraint(:organization)
   end
 
+  defp associate_categories(changeset, %{"categories_ids" => []}) do
+    do_associate_categories(changeset, [])
+  end
+  defp associate_categories(changeset, %{"categories_ids" => ids}) do
+    categories =
+      ids
+      |> Enum.join(",")
+      |> coalesce_id_string()
+      |> find_categories()
+
+    changeset |> put_assoc(:categories, categories)
+  end
+  defp associate_categories(changeset, _) do
+    do_associate_categories(changeset, [])
+  end
+
+  defp do_associate_categories(changeset, _) do
+    changeset |> put_assoc(:categories, [])
+  end
+
+  defp find_categories(ids) do
+    Category
+    |> where([object], object.id in ^ids)
+    |> Repo.all()
+  end
+
+  defp associate_skills(changeset, %{"skills_ids" => []}) do
+    do_associate_skills(changeset, [])
+  end
+  defp associate_skills(changeset, %{"skills_ids" => ids}) do
+    skills =
+      ids
+      |> Enum.join(",")
+      |> coalesce_id_string()
+      |> find_skills()
+
+    changeset |> put_assoc(:skills, skills)
+  end
+  defp associate_skills(changeset, _) do
+    do_associate_skills(changeset, [])
+  end
+
+  defp do_associate_skills(changeset, _) do
+    changeset |> put_assoc(:skills, [])
+  end
```

Comment:
We could save us a clause, and a `do_associate` helper, by using these:

```Elixir
defp associate_categories(changeset, %{"categories_ids" => ids}) when is_binary(ids) do
  categories =
    ids |> Enum.join(",") |> coalesce_id_string() |> find_categories()
  changeset |> put_assoc(:categories, categories)
end
defp associate_categories(changeset, _), do: changeset |> put_assoc(:categories, [])
```

Reply:
Note: I'm seeing some inconsistencies with `coalesce_ids` string and some of our tests.
If we're supposed to be getting a list here, instead of a csv `ids`string, then the first clause would guard with

```Elixir
when is_list(ids) when length(ids) > 0 do
```